Add support for using topologyKey=karpenter.sh/capacity-type in TopologySpreadConstraint

rajakshay opened the pull request:
**1. Issue, if available:**
https://github.com/aws/karpenter/issues/1557

**2. Description of changes:**
* Allows using karpenter.sh/capacity-type as topologyKey in TopologySpreadConstraint

**3. How was this change tested?**
* Tested on EKS using the instructions here: https://karpenter.sh/v0.7.3/development-guide/ 
* Karpenter was able to spin up the correct `capacity-type` based on the PodSpec to respect the distribution. Tested for distribution of pods across on-demand and spot EC2 instances. 

* podSpec used for testing topologySpreadConstraint:
```
topologySpreadConstraints:
      - labelSelector:
          matchLabels:
            app: <app-name>
        maxSkew: 1
        topologyKey: karpenter.sh/capacity-type
        whenUnsatisfiable: DoNotSchedule
```

* Karpenter provisioner specified:
```
  requirements:
  - key: karpenter.sh/capacity-type
    operator: In
    values:
    - spot
    - on-demand
```

* Karpenter was successfully spinning up nodes across on-demand and spot as the workload instances was scaled up. 

**4. Does this change impact docs?**
- [x] Yes, PR includes docs updates
- [ ] Yes, issue opened: *link to issue*
- [ ] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

tzneal reviewed (changes requested):
Can you add some tests to `scheduling/suite_test.go`?  There's a section there they'll fit in at: `Describe("Topology",`

ellistarn commented:
I just noticed:
> [ x] No

This change does impact docs.

rajakshay commented:
> Can you add some tests to `scheduling/suite_test.go`? There's a section there they'll fit in at: `Describe("Topology",`

Thank you for the quick response and pointer for adding tests. Im working on adding some tests. I'll push an update in a few hours.

rothgar commented:
Will this change allow me to say I want an even spread between spot and on-demand? Am I able to specify something like 20% spot and 80% on demand with this? It doesn't look like it but I'm not sure the intended use case.

ellistarn commented:
> Will this change allow me to say I want an even spread between spot and on-demand? Am I able to specify something like 20% spot and 80% on demand with this? It doesn't look like it but I'm not sure the intended use case.

I'd love to extend upstream topology to do percentage based domains. Right now, the spec implies an even split across domains (e.g. 50/50 here).

tzneal commented:
@rajakshay Sorry, I was trying to add some docs as a pull-request to your PR but it ended up just adding them on as a new commit on this PR.  Feel free to rebase and squash that into your commit if you would like.

rajakshay commented:
> @rajakshay Sorry, I was trying to add some docs as a pull-request to your PR but it ended up just adding them on as a new commit on this PR. Feel free to rebase and squash that into your commit if you would like.

@tzneal Thank you for adding the docs. I updated an existing example as well, and squashed all commits together. This is ready for review.

tzneal reviewed (approved):
LGTM, I pulled this branch and verified that capacity-type spread is working on my cluster.
